pyspectral/rayleigh.py

```diff
@@ -262,80 +263,39 @@ def get_reflectance(self, sun_zenith, sat_zenith, azidiff, bandname, redband=Non
+        # version of the algorithm but return numpy arrays back
+        compute = HAVE_DASK and not isinstance(sun_zenith, Array)
+
+        rayleigh_refl = _get_wavelength_adjusted_lut_rayleigh_reflectance(self.reflectance_lut_filename, wvl)
+        if rayleigh_refl is None:
+            LOG.warning("Effective wavelength for band %s outside 400-800 nm range!",
+                        str(bandname))
+            LOG.info("Setting the rayleigh/aerosol reflectance contribution to zero!")
+            return _get_zeroed_result(
+                sun_zenith.shape, compute,
+                chunks=sun_zenith.chunks if redband is None else redband.chunks)
+        res = map_blocks(self._interp_rayleigh_refl_by_angles,
+                         sun_zenith, sat_zenith, azidiff, rayleigh_refl,
+                         self.reflectance_lut_filename,
+                         meta=np.array((), dtype=rayleigh_refl.dtype),
+                         dtype=rayleigh_refl.dtype,
+                         chunks=getattr(azidiff, "chunks", None))
 
-            return zeros(shape)
-
-        idx = np.searchsorted(wvl_coord, wvl)
-        wvl1 = wvl_coord[idx - 1]
-        wvl2 = wvl_coord[idx]
-
-        fac = (wvl2 - wvl) / (wvl2 - wvl1)
-        raylwvl = fac * rayl[idx - 1, :, :, :] + (1 - fac) * rayl[idx, :, :, :]
-        tic = time.time()
-
-        smin = [sunz_sec_coord[0], azid_coord[0], satz_sec_coord[0]]
-        smax = [sunz_sec_coord[-1], azid_coord[-1], satz_sec_coord[-1]]
-        orders = [
-            len(sunz_sec_coord), len(azid_coord), len(satz_sec_coord)]
-        f_3d_grid = atleast_2d(raylwvl.ravel())
-
-        if HAVE_DASK and isinstance(smin[0], Array):
-            # compute all of these at the same time before passing to the interpolator
-            # otherwise they are computed separately
-            smin, smax, orders, f_3d_grid = da.compute(smin, smax, orders, f_3d_grid)
-        minterp = MultilinearInterpolator(smin, smax, orders)
-        minterp.set_values(f_3d_grid)
-
-        if HAVE_DASK:
-            ipn = map_blocks(self._do_interp, minterp, sunzsec, azidiff,
-                             satzsec, dtype=raylwvl.dtype, chunks=azidiff.chunks)
-        else:
-            ipn = self._do_interp(minterp, sunzsec, azidiff, satzsec)
-
-        LOG.debug("Time - Interpolation: {0:f}".format(time.time() - tic))
-
-        ipn *= 100
-        res = ipn
         if redband is not None:
-            res = where(redband < 20., res,
-                        (1 - (redband - 20) / 80) * res)
+            res = map_blocks(self._correct_red_by_interpolated_rayleigh_refl,
+                             redband, res)
 
-        res = clip(nan_to_num(res), 0, 100)
+        res = np.clip(np.nan_to_num(res), 0, 100)
         if compute:
             res = res.compute()
-
         return res
 
+    @staticmethod
+    def _correct_red_by_interpolated_rayleigh_refl(redband, rayleigh_refl):
```

Comment:
Maybe reconsider the name of this function?
I didn't understand it in relation to what its doing.
The purpose here is to relax the rayleigh correction in cloudy areas. We use the red band reflection as a proxy for cloud presence. The modelled correction strictly applied only in a cloudfree atmosphere. 
It is hard with appropriate names: Maybe `_relax_rayleigh_reflection_correction_term_where_cloudy`? Or along those lines?

Reply:
Ooohhh I think I have always just misunderstood what the point of the `redband` parameter was in this code. I thought this was applying the rayleigh correction to the red band but this is obviously wrong now that I've thought about it for two seconds. I don't mind your suggested long name. I may have just used `_adjust_rayleigh_with_red_band_clouds` or something like that. I can use your name.

One question that this reminds me of though: what is the reflectance variable in the LUT file? It is called `reflectance` in the file, but then called `rayleigh` later and then called `raywl` or something later. So I just started calling it "rayleigh reflectance", but is it a reflectance? Is it a factor? What is it?

Reply:
@djhoese It is a reflectance. The data in the LUT file is the simulated clear-sky reflectance with a dark background (none-reflective surface). So, that is mainly the Rayleigh scattering contribution coming from the atmosphere - but it is not exclusively Rayleigh scattering, as the simulations of this signal includes aerosol absorption. Thus, the value from the LUT should be subtracted from the observed reflectance to obtain the (atmosphere) corrected reflectance.

Don't know if @rolle-at-work wants to add something here?